File: Algorithms_python/5_Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        start = 0
        Max = 1
        for j in range(n):
            for i in range(j + 1):
                if j - i < 2:
                    dp[i][j] = s[i] == s[j]
                else:
                    dp[i][j] = dp[i + 1][j - 1] and s[i] == s[j]
                if dp[i][j] and Max < j -i + 1:
                    start = i
                    Max = j - i + 1
        return s[start:start + Max]

'''
动态规划法。
设 dp[i][j] 表示字符串 s[i..j] 是否为回文串。
当 j - i < 2，即字符串长度为 2 时，只要 s[i] == s[j]，dp[i][j] 就为 true。
当 j - i >= 2，dp[i][j] = dp[i + 1][j - 1] && s[i] == s[j]。
'''

File: Algorithms_C#/5_Longest_Palindromic_Substring.cs
public class Solution {
    public int[] expandAroundCenter(string s, int left, int right){
        while (left >= 0 && right < s.Length && s[left] == s[right]) {
            --left;
            ++right;
        }
        return new int[] {left + 1, right - 1};
    }
    public string LongestPalindrome(string s) {
        int start = 0, end = 0;
        for (int i = 0; i < s.Length; ++i) {
            int[] odd = expandAroundCenter(s, i, i);
            int[] even = expandAroundCenter(s, i, i + 1);
            if (odd[1] - odd[0] > end - start) {
                start = odd[0];
                end = odd[1];
            }
            if (even[1] - even[0] > end - start) {
                start = even[0];
                end = even[1];
            }
        }
        return s.Substring(start, end - start + 1);
    }
}